src/ucm/bistro/bistro_x86_64.c

```diff
@@ -14,19 +14,153 @@
 #if defined(__x86_64__)
 
 #include <sys/mman.h>
-#include <dlfcn.h>
 #include <string.h>
 #include <stdlib.h>
+#include <pthread.h>
 
 #include <ucm/bistro/bistro.h>
 #include <ucm/bistro/bistro_int.h>
 #include <ucm/util/sys.h>
 #include <ucs/sys/math.h>
-#include <ucs/arch/cpu.h>
-#include <ucs/debug/assert.h>
 
 
-ucs_status_t ucm_bistro_patch(const char *symbol, void *hook,
+typedef struct {
+    void    *jmp_address;
+    char    code[];
+} ucm_bistro_orig_func_t;
+
+typedef struct {
+    uint8_t  opcode; /* 0xff */
+    uint8_t  modrm;  /* 0x25 */
+    int32_t  displ;
+} UCS_S_PACKED ucm_bistro_jmp_indirect_t;
+
+
+/*
+ * Find the minimal length of initial instructions in the function which can be
+ * safely executed from any memory location.
+ * Uses a very simplified disassembler which supports only the typical
+ * instructions found in function prologue.
+ */
+static size_t ucm_bistro_find_movable_prefix(const void *func, size_t min_length)
+{
+    size_t offset, prev_offset;
+    uint8_t rex, opcode, modrm;
+
+    offset = 0;
+    while (offset < min_length) {
+        prev_offset = offset;
+        opcode      = *(uint8_t*)UCS_PTR_BYTE_OFFSET(func, offset++);
+
+        /* check for REX prefix */
+        if ((opcode & 0xF0) == 0x40) {
+            rex    = opcode;
+            opcode = *(uint8_t*)UCS_PTR_BYTE_OFFSET(func, offset++);
+        } else {
+            rex = 0;
+        }
+
+        /* check the opcode */
+        if (((rex == 0) || rex == 0x41) && ((opcode & 0xF0) == 0x50)) {
```

Comment:
```((rex == 0) || rex == 0x41)``` is this correct?

Reply:
rex == 0 is pushing IA32 register and rex==0x41 is pushing rN register